### PR TITLE
RSE-133: Fix: Notifications payload format are not imported properly by the API and the UI

### PR DESCRIPTION
**Pull request by avelasquezr**
## BugFix: Notifications payload format are not imported properly by the API and the UI

Fix to modify the LinkedHashMap mapping and insert a single map to the array. Related to https://github.com/rundeck/rundeck/issues/6711

#### Considering the following YAML:
``` yaml
# YAML
...
notification:
  onfailure:
    format: json
    method: post
    urls: http://rundeck-server:4440/api/36/webhook/TOKEN#webhook
  onsuccess:
    format: json
    method: post
    urls: http://rundeck-server:4440/api/36/webhook/TOKEN#webhook
```
### The Problem:
When the yaml is identified as coming in Map format (without "-", this symbol reference a list type), each map is inserted into the list:
code reference:
```
iteration {
   map.put(key,value)
   array.add(map)
}
```
### The Fix:
The mapping of the LinkedHashMap is modified to insert a single map to the array.
code reference:
```
iteration {
   map.put(key,value)
}
array.add(map)
```
Now list types and map types are accepted
``` yaml
# YAML
...
notification:
  onfailure:
    - format: json
       method: post
       urls: http://rundeck-server:4440/api/36/webhook/TOKEN#webhook
    - format: json
       method: post
       urls: http://rundeck-server:4440/api/36/webhook/TOKEN#webhook
  onsuccess:
    format: json
    method: post
    urls: http://rundeck-server:4440/api/36/webhook/TOKEN#webhook
```

### CODE REVIEW

File: rundeckapp/src/test/groovy/org/rundeck/app/components/JobYAMLFormatSpec.groovy [hunk 1]
```diff
@@ -264,6 +264,93 @@ class JobYAMLFormatSpec extends Specification {
         result[0].notification['onsuccess'].findAll{ it['urls'] != null }.size() == 2
     }
 
+    @Unroll
+    def "should return a notification map webhook url notifs"() {
```
Comment (L2JE) [inline rundeckapp/src/test/groovy/org/rundeck/app/components/JobYAMLFormatSpec.groovy:268]: The decode method always returns list of notifications for each trigger. Example of better name: `"should return a notif list with one webhook notif"`

**Review by L2JE — CHANGES_REQUESTED** [review]
It seems to not be working on the case of having two different notifications in the same trigger without the list style:
```
  notification:
    onstart:
      format: json
      httpMethod: post
      urls: http://localhost:4440/project
      email:
        attachLog: 'true'
        attachLogInFile: true
        recipients: tom@example.com
        subject: JOB-STARTED
```

Link to official docs example:
https://docs.rundeck.com/docs/manual/document-format-reference/job-yaml-v12.html#another-valid-format

File: rundeckapp/src/test/groovy/org/rundeck/app/components/JobYAMLFormatSpec.groovy (hunk 1 above)
Comment (avelasquezr) [reply rundeckapp/src/test/groovy/org/rundeck/app/components/JobYAMLFormatSpec.groovy]: Test - renamed

File: rundeckapp/src/test/groovy/org/rundeck/app/components/JobYAMLFormatSpec.groovy [hunk 2]
```diff
@@ -264,6 +264,166 @@ class JobYAMLFormatSpec extends Specification {
...
+                "          attachLog: true\n" +
+                "          attachLogInline: true\n" +
+                "          recipients: 2@gmail.com\n" +
+                "          subject: RD-SUCCESS\n" +
+                "      - format: xml\n" +
+                "        httpMethod: get\n" +
+                "        urls: http://localhost:4440/project\n" +
+                "      - format: json\n" +
+                "        httpMethod: post\n" +
+                "        urls: http://localhost:4440/project\n" +
```
Comment (L2JE) [inline rundeckapp/src/test/groovy/org/rundeck/app/components/JobYAMLFormatSpec.groovy:327]: one or two elements should be enough to represent the list for this test case

File: rundeckapp/src/test/groovy/org/rundeck/app/components/JobYAMLFormatSpec.groovy [hunk 3]
```diff
@@ -264,6 +264,166 @@ class JobYAMLFormatSpec extends Specification {
...
+        def sut = new JobYAMLFormat()
+        when:
+        def result = sut.decode(new StringReader(input))
+        then:
+        result[0].notification['onsuccess'].size() == 1
+        result[0].notification['onsuccess'].findAll{ it['urls'] != null }.size() == 1
+    }
+
+    @Unroll
+    def "should return a notification list of map and single map with email and webhook notifs"() {
```
Comment (L2JE) [inline rundeckapp/src/test/groovy/org/rundeck/app/components/JobYAMLFormatSpec.groovy:301]: This name sounds confusing.

This test seems to be trying to check that either a list of notifications or a notification map will result on a list of notifications for each trigger always. So I think a more suitable name (thinking on the testcase and not on the exact code) for this test would be something like:
`should return list of notifications for each trigger when some triggers have notifs in yaml list and some not in a list`

File: rundeckapp/src/test/groovy/org/rundeck/app/components/JobYAMLFormatSpec.groovy [hunk 4]
```diff
@@ -264,6 +264,166 @@ class JobYAMLFormatSpec extends Specification {
...
+        result[0].notification['onsuccess'].size() == 4
+        result[0].notification['onsuccess'].findAll{ it['email'] != null }.size() == 2
+        result[0].notification['onsuccess'].findAll{ it['urls'] != null }.size() == 2
+
+        result[0].notification['onfailure'].size() == 1
+        result[0].notification['onfailure'].findAll{ it['urls'] != null }.size() == 1
+    }
+
+    @Unroll
+    def "should return a notification list of map and single map with email, plugin and webhook notify"() {
```
Comment (L2JE) [inline rundeckapp/src/test/groovy/org/rundeck/app/components/JobYAMLFormatSpec.groovy:355]: Same as previous one, confusing name, thinking from the use case point of view, a better name for this case could be:
`should return list of notifications for each trigger when all triggers have notifs not in yaml lists` or 
`should return list of notifications for each trigger when all triggers have notifs in yaml maps`

**Review by L2JE — CHANGES_REQUESTED** [review]

**Review by L2JE — APPROVED** [review]
LGTM

**Pull request merged**